addons/web/controllers/main.py

```diff
@@ -924,6 +924,21 @@ def web_login(self, redirect=None, **kw):
         response.headers['X-Frame-Options'] = 'DENY'
         return response
 
+    # sid changes once connected, therefore csrf changes too.
+    @http.route('/web/accept_scope', type='http', auth="user", methods=['GET'], website=True)
+    def accept_scope(self, **values):
+        return request.render('web.app_auth', values)
+
+    @http.route('/web/oauth', type='http', auth="user", methods=['POST'])
+    def web_oauth(self, redirect=None, scope=None, do="refuse", **kw):
```

Comment:
Calling it `oauth` seems misleading as this is only a very small part of a full oauth flow. Also might be a good idea to have a dedicated secret / signature / ... (aside from the csrf) to avoid the case where an XSS would allow generating tokens on the fly?

And `redirect` has no reason to be optional, as both paths will misbehave or break if it's not provided.

Reply:
> Also might be a good idea to have a dedicated secret / signature / ... (aside from the csrf) to avoid the case where an XSS would allow generating tokens on the fly

I'm not sure I understand that part. Do you mean a code that would be embedded in the page, hence leaked...

Reply:
>  Do you mean a code that would be embedded in the page, hence leaked...

Doesn't matter, the point is to ensure the user has gone through the proper flow. The CSRF token is not per-endpoint (as the name indicates it's only intended to protect against *cross-site* request forgery), so it's possible to open a page, get the CSRF token, then use that CSRF token against other endpoints.

Meaning if we happen to have an XSS on a page with a form (and thus a CSRF token), an attacker could capture that CSRF token, then use it against this endpoint and generate oauth tokens, I think anyway.